kaldb/src/main/java/com/slack/kaldb/chunk/ReadOnlyChunkImpl.java

```diff
@@ -27,115 +37,169 @@
+    // this.chunkinfo = ?
 
-  @Override
-  public void setReadOnly(boolean readOnly) {
-    if (!readOnly) {
-      throw new UnsupportedOperationException("ReadOnly chunk can't be set to write.");
-    }
+    setChunkMetadataState(Metadata.CacheSlotState.LIVE);
   }
 
-  @Override
-  public void preSnapshot() {
-    LOG.info("Finished pre-snapshot for RO chunk {}", chunkInfo);
+  private void handleChunkEviction() throws Exception {
+    setChunkMetadataState(Metadata.CacheSlotState.EVICTING);
+
+    chunkInfo = null;
+    logSearcher = null;
+
+    setChunkMetadataState(Metadata.CacheSlotState.FREE);
   }
 
-  @Override
-  public boolean snapshotToS3(String bucket, String prefix, S3BlobFs s3BlobFs) {
-    LOG.info("Failed snapshot to S3 for RO chunk {}", chunkInfo);
-    throw new UnsupportedOperationException("ReadOnly chunk can't be snapshotted.");
+  @VisibleForTesting
+  public void setChunkMetadataState(Metadata.CacheSlotState newChunkState) throws Exception {
+    CacheSlotMetadata chunkMetadata = cacheSlotMetadataStore.getCached().get(0);
+    CacheSlotMetadata updatedChunkMetadata =
+        new CacheSlotMetadata(
+            chunkMetadata.name,
+            newChunkState,
+            newChunkState.equals(Metadata.CacheSlotState.FREE) ? "" : chunkMetadata.replicaId,
+            Instant.now().toEpochMilli());
+    cacheSlotMetadataStore.update(updatedChunkMetadata).get(5, TimeUnit.SECONDS);
   }
 
-  @Override
-  public void postSnapshot() {
-    LOG.info("Finished post-snapshot for RO chunk {}", chunkInfo);
+  @VisibleForTesting
+  public Metadata.CacheSlotState getChunkMetadataState() {
+    if (cacheSlotMetadataStore.getCached().isEmpty()) {
+      return null;
+    }
+    return cacheSlotMetadataStore.getCached().get(0).cacheSlotState;
   }
 
-  /** Deletes the log store data from local disk. Should be called after close(). */
   @Override
-  public void cleanup() {
-    // TODO: Implement chunk state cleanup
-    throw new UnsupportedOperationException("To be implemented.");
+  public ChunkInfo info() {
+    return chunkInfo;
   }
 
   @Override
-  @VisibleForTesting
-  public LogIndexSearcher<T> getLogSearcher() {
-    return logSearcher;
+  public boolean containsDataInTimeRange(long startTs, long endTs) {
+    if (chunkInfo != null) {
+      return chunkInfo.containsDataInTimeRange(startTs, endTs);
+    }
+    return false;
   }
 
   @Override
-  @VisibleForTesting
-  public void setLogSearcher(LogIndexSearcher<T> logSearcher) {
-    this.logSearcher = logSearcher;
+  public void close() throws IOException {
+    if (logSearcher != null) {
+      logSearcher.close();
+    }
+
+    executorService.shutdown();
+    cacheSlotMetadataStore.close();
+
+    LOG.info("Closed chunk {}", chunkId);
+    cleanup();
   }
 
-  @Override
-  public boolean isReadOnly() {
-    return readOnly;
+  /** Deletes the log store data from local disk. */
+  public void cleanup() {
+    // TODO: Implement chunk state cleanup
```

Comment:
I think we should implement this when we add the caching store.

Reply:
This was addressed as part of a fix above - https://github.com/slackhq/kaldb/pull/122#discussion_r715809724